File: java/src/main/java/it/giannibombelli/wsc2026/payment/api/PaymentApi.java
package it.giannibombelli.wsc2026.payment.api;

import io.javalin.config.JavalinConfig;
import io.javalin.http.Context;
import io.javalin.openapi.*;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.StartTransaction;
import it.giannibombelli.wsc2026.payment.application.query.PaymentDetails;
import it.giannibombelli.wsc2026.payment.application.query.PaymentFinder;
import it.giannibombelli.wsc2026.payment.application.services.PaymentProcessing;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;

import java.util.UUID;

import static io.javalin.apibuilder.ApiBuilder.get;
import static io.javalin.apibuilder.ApiBuilder.post;

public final class PaymentApi {
    private final PaymentFinder paymentFinder;
    private final PaymentProcessing paymentProcessing;

    public PaymentApi(PaymentFinder paymentFinder, PaymentProcessing paymentProcessing) {
        this.paymentFinder = paymentFinder;
        this.paymentProcessing = paymentProcessing;
    }

    public void configure(JavalinConfig config) {
        config.routes.apiBuilder(() -> {
            get("/payments/{id}", this::getById);
            post("/payments/{id}/transactions", this::startTransaction);
        });
    }

    @OpenApi(
        path = "/payments/{id}/transactions",
        methods = HttpMethod.POST,
        summary = "Start a payment transaction (asynchronous)",
        tags = {"Payment"},
        pathParams = {@OpenApiParam(name = "id", required = true)},
        requestBody = @OpenApiRequestBody(content = {@OpenApiContent(from = StartTransactionRequest.class)}),
        responses = {
            @OpenApiResponse(status = "202", content = {@OpenApiContent(from = TransactionResponse.class)}),
            @OpenApiResponse(status = "400")
        }
    )
    private void startTransaction(Context ctx) {
        PaymentId paymentId = parsePaymentId(ctx);
        if (paymentId == null) {
            return;
        }

        StartTransactionRequest request = parseStartTransactionRequest(ctx);
        if (request == null) {
            return;
        }

        if (request.provider() == null || request.provider().isBlank()) {
            ctx.status(400).result("provider is required");
            return;
        }
        if (request.amount() == null) {
            ctx.status(400).result("amount is required");
            return;
        }

        Money amount;
        try {
            amount = new Money(request.amount());
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        StartTransaction cmd;
        try {
            ProviderReference providerReference = request.providerReference() == null ? null : new ProviderReference(request.providerReference());
            cmd = new StartTransaction(paymentId, Provider.fromLabel(request.provider()), providerReference, amount, Timestamp.now());
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        if (paymentFinder.findDetailsById(paymentId).isEmpty()) {
            ctx.status(404);
            return;
        }

        TransactionStarted started;
        try {
            started = paymentProcessing.invoke(cmd);
        } catch (IllegalArgumentException | IllegalStateException ex) {
            ctx.status(400).result(ex.getMessage());
            return;
        }

        TransactionResponse response = new TransactionResponse(
            started.transactionId().value(),
            started.provider().label(),
            cmd.providerReference() == null ? null : cmd.providerReference().value(),
            started.amount().value(),
            "STARTED",
            cmd.startedAt().value(),
            null
        );
        ctx.status(202).json(response);
    }

    @OpenApi(
        path = "/payments/{id}",
        methods = HttpMethod.GET,
        summary = "Get a payment by id",
        tags = {"Payment"},
        pathParams = {@OpenApiParam(name = "id", required = true)},
        responses = {
            @OpenApiResponse(status = "200", content = {@OpenApiContent(from = PaymentDetailsResponse.class)}),
            @OpenApiResponse(status = "400"),
            @OpenApiResponse(status = "404")
        }
    )
    private void getById(Context ctx) {
        PaymentId id = parsePaymentId(ctx);
        if (id == null) {
            return;
        }

        var paymentOpt = paymentFinder.findDetailsById(id);
        if (paymentOpt.isEmpty()) {
            ctx.status(404);
            return;
        }

        ctx.status(200).json(PaymentDetailsResponse.from(paymentOpt.get()));
    }

    private PaymentId parsePaymentId(Context ctx) {
        String idParam = ctx.pathParam("id");
        try {
            return new PaymentId(UUID.fromString(idParam));
        } catch (IllegalArgumentException ex) {
            ctx.status(400).result("Invalid payment id format");
            return null;
        }
    }

    private StartTransactionRequest parseStartTransactionRequest(Context ctx) {
        try {
            return ctx.bodyAsClass(StartTransactionRequest.class);
        } catch (Exception ex) {
            ctx.status(400).result("request body is required");
            return null;
        }
    }

    private static TransactionResponse toTransactionResponse(PaymentDetails.TransactionDetail detail) {
        return new TransactionResponse(
            detail.id(),
            detail.provider().label(),
            detail.providerReference(),
            detail.amount().value(),
            detail.status().name(),
            detail.startedAt().value(),
            detail.completedAt() == null ? null : detail.completedAt().value()
        );
    }
}

File: typescript/src/payment/api/paymentApi.ts
// API pubbliche del Payment Bounded Context.

import express, { type Express, type Request, type Response } from 'express';
import { getErrorMessage } from '@/common/api/errorMessage.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { StartTransaction, startTransaction } from '../application/commands/startTransaction.js';
import { PaymentFinder } from '../application/query/paymentFinder.js';
import { PaymentProcessing } from '../application/services/paymentProcessing.js';
import { PaymentId } from '../domain/payment/paymentId.js';
import { PaymentNotFoundException } from '../domain/payment/paymentNotFoundException.js';
import { providerFromLabel } from '../domain/payment/provider.js';
import { ProviderReference } from '../domain/payment/providerReference.js';
import { parseStartTransactionRequest } from './startTransactionRequest.js';
import { toPaymentDetailsResponse } from './paymentDetailsResponse.js';
import { toTransactionResponse } from './transactionResponse.js';
import { requireDependency } from '@/common/utils/requireDependency.js';
import type { WebApi } from '@/common/module/applicationModule.js';

export class PaymentApi implements WebApi {
  private readonly _paymentFinder: PaymentFinder;
  private readonly _paymentProcessing: PaymentProcessing;

  constructor(paymentFinder: PaymentFinder, paymentProcessing: PaymentProcessing) {
    requireDependency(paymentFinder, "paymentFinder");
    requireDependency(paymentProcessing, "paymentProcessing");
    this._paymentFinder = paymentFinder;
    this._paymentProcessing = paymentProcessing;
  }

  configure(app: Express): void {
    const router = express.Router();
    router.use(express.json());

    router.get('/payments/:id', (req, res) => this.getById(req, res));
    router.post('/payments/:id/transactions', (req, res) => this.startTransaction(req, res));

    app.use(router);
  }

  /**
   * @swagger
   * /payments/{id}/transactions:
   *   post:
   *     summary: Avvia una transazione di pagamento (asincrona)
   *     tags: [Payment]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: string
   *           format: uuid
   *     requestBody:
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/StartTransactionRequest'
   *     responses:
   *       202:
   *         description: Transazione avviata
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/TransactionResponse'
   *       400:
   *         description: Richiesta non valida
   *       404:
   *         description: Pagamento non trovato
   */
  private startTransaction(req: Request, res: Response): void {
    const paymentId = this.parsePaymentId(req, res);
    if (paymentId === null) {
      return;
    }

    let request;
    try {
      request = parseStartTransactionRequest(req.body);
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    let amount: Money;
    try {
      amount = new Money(Number(request.amount));
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'invalid amount'));
      return;
    }

    let cmd: StartTransaction;
    try {
      const providerReference = request.providerReference === null ? null : new ProviderReference(Uuid.fromString(request.providerReference));
      cmd = startTransaction(
        paymentId,
        providerFromLabel(request.provider),
        providerReference,
        amount,
        Timestamp.now()
      );
    } catch (error) {
      res.status(400).send(getErrorMessage(error, 'invalid request'));
      return;
    }

    let started;
    try {
      started = this._paymentProcessing.invoke(cmd);
    } catch (error) {
      if (error instanceof PaymentNotFoundException) {
        res.status(404).send();
        return;
      }
      res.status(400).send(getErrorMessage(error, 'bad request'));
      return;
    }

    const response = toTransactionResponse({
      id: started.transactionId.value.value,
      provider: started.provider,
      providerReference: cmd.providerReference === null ? null : cmd.providerReference.value.value,
      amount: started.amount,
      status: 'STARTED',
      startedAt: cmd.startedAt,
      completedAt: null,
    });
    res.status(202).json(response);
  }

  /**
   * @swagger
   * /payments/{id}:
   *   get:
   *     summary: Recupera il dettaglio di un pagamento
   *     tags: [Payment]
   *     parameters:
   *       - in: path
   *         name: id
   *         required: true
   *         schema:
   *           type: string
   *           format: uuid
   *     responses:
   *       200:
   *         description: Pagamento trovato
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/PaymentDetailsResponse'
   *       400:
   *         description: ID non valido
   *       404:
   *         description: Pagamento non trovato
   */
  private getById(req: Request, res: Response): void {
    const id = this.parsePaymentId(req, res);
    if (id === null) {
      return;
    }

    const payment = this._paymentFinder.findDetailsById(id);
    if (payment === null) {
      res.status(404).send();
      return;
    }

    res.status(200).json(toPaymentDetailsResponse(payment));
  }

  private parsePaymentId(req: Request, res: Response): PaymentId | null {
    const idParam = req.params.id;
    if (typeof idParam !== 'string') {
      res.status(400).send('Invalid payment id format');
      return null;
    }
    try {
      return new PaymentId(Uuid.fromString(idParam));
    } catch (_error) {
      res.status(400).send('Invalid payment id format');
      return null;
    }
  }
}
